PublicMapping/district-builder-2
# User registration

## Overview

This pull request adds a 2 step user registration system with email verification

### Demo

```
$ http http://localhost:3005/auth/email/register email=mmaurizi@azavea.com password=mike12345 name="Mike M"
HTTP/1.1 201 Created                                                                                                                                                                    
Connection: keep-alive                                                                                                                                                                  
Content-Length: 7
Content-Type: text/html; charset=utf-8                                                                                                                                                  
Date: Thu, 02 Apr 2020 15:39:44 GMT                                                                                                                                                     
ETag: W/"7-gFw9aUPkrb4JBgwx0K+SLEMaCno"                                                                                                                                                 
X-Powered-By: Express                                                                                                                                                                   
                                                                                                                                                                                        
SUCCESS

$ http POST http://localhost:3005/auth/email/resend-verification/mmaurizi@azavea.com
HTTP/1.1 201 Created
Connection: keep-alive
Content-Length: 7
Content-Type: text/html; charset=utf-8
Date: Thu, 02 Apr 2020 16:13:31 GMT
ETag: W/"7-gFw9aUPkrb4JBgwx0K+SLEMaCno"
X-Powered-By: Express

SUCCESS

$ http POST http://localhost:3005/auth/email/verify/Iku9ORZhSplZzKRSyWoo7vDJn9U
HTTP/1.1 201 Created
Connection: keep-alive
Content-Length: 114
Content-Type: application/json; charset=utf-8
Date: Thu, 02 Apr 2020 16:13:50 GMT
ETag: W/"72-kDki5zX15WVb9OY36DyWP6gSu3M"
X-Powered-By: Express

{
    "email": "mmaurizi@azavea.com",
    "id": "0614a225-8d2b-4c71-acdc-d66b4869b470",
    "isEmailVerified": true,
    "name": "Mike M"
}

$ http POST http://localhost:3005/auth/email/login email=mmaurizi@azavea.com password=mike12345
HTTP/1.1 201 Created
Connection: keep-alive
Content-Length: 114
Content-Type: application/json; charset=utf-8
Date: Thu, 02 Apr 2020 16:41:44 GMT
ETag: W/"72-kDki5zX15WVb9OY36DyWP6gSu3M"
X-Powered-By: Express

{
    "email": "mmaurizi@azavea.com",
    "id": "0614a225-8d2b-4c71-acdc-d66b4869b470",
    "isEmailVerified": true,
    "name": "Mike M"
}

$ http POST http://localhost:3005/auth/email/login email=mmaurizi@azavea.com password=mike54321
HTTP/1.1 400 Bad Request
Connection: keep-alive
Content-Length: 74
Content-Type: application/json; charset=utf-8
Date: Thu, 02 Apr 2020 16:41:52 GMT
ETag: W/"4a-J+TcAdnMDmQd2xHUXvqyIXRNQbg"
X-Powered-By: Express

{
    "error": "Invalid password",
    "message": "INVALID_PASSWORD",
    "statusCode": 400
}


```

### Notes

 - I haven't disabled `login` for users whose emails are not verified because I didn't know if we wanted to restrict login entirely or merely restrict access in the application to features that need email.
 - I'm deferring password validation to a future issue, hopefully I can find something that has an equivalent React UI component to validate during registration in #52 
 - We need to add issues Forgot password & change password APIs & frontend pages, we missed those in our initial planning
 - We should add an issue to update the email registration copy / style it.
 - I didn't 1-for-1 copy much, but I used this as a reference work quite a bit: https://github.com/marcomelilli/nestjs-email-authentication/

## Testing Instructions

- `scripts/update`
- Create a user using the `/auth/email/register`endpoint
- Login using `auth/email/login`
- Verify that you see email contents printed to the console running `scripts/server`
- Test resending an email with `/auth/email/resend-verification/:email`
- Using the token in the email printed to the console, verify your user using `auth/email/verify/:token` 

Closes #49
